File: server/controllers/userController.js
const User = require('../models/User');
const PollutionReport = require('../models/PollutionReport');

// @route   GET /api/users/leaderboard
exports.getLeaderboard = async (req, res) => {
    try {
        const { limit = 20 } = req.query;

        const users = await User.find()
            .select('name avatar_url contribution_score reports_count verified_reports badges created_at')
            .sort({ contribution_score: -1 })
            .limit(parseInt(limit))
            .lean();

        const leaderboard = users.map((user, index) => ({
            rank: index + 1,
            ...user,
            id: user._id,
            badges: user.badges || []
        }));

        res.json({ success: true, data: leaderboard });
    } catch (error) {
        console.error('Leaderboard error:', error);
        res.status(500).json({ success: false, message: 'Server error fetching leaderboard' });
    }
};

// @route   GET /api/users/:id/profile
exports.getUserProfile = async (req, res) => {
    try {
        const { id } = req.params;

        const user = await User.findById(id)
            .select('name avatar_url contribution_score reports_count verified_reports badges created_at')
            .lean();

        if (!user) {
            return res.status(404).json({ success: false, message: 'User not found' });
        }

        const reports = await PollutionReport.find({ user_id: id })
            .select('title category severity ai_verified status upvotes downvotes created_at latitude longitude')
            .sort({ created_at: -1 })
            .limit(10)
            .lean();

        res.json({
            success: true,
            data: {
                user: { ...user, id: user._id, badges: user.badges || [] },
                recent_reports: reports
            }
        });
    } catch (error) {
        console.error('User profile error:', error);
        res.status(500).json({ success: false, message: 'Server error fetching profile' });
    }
};

// @route   GET /api/users/stats
exports.getPlatformStats = async (req, res) => {
    try {
        const [usersCount, reportsCount, verifiedCount, citiesAgg] = await Promise.all([
            User.countDocuments(),
            PollutionReport.countDocuments(),
            PollutionReport.countDocuments({ ai_verified: true }),
            PollutionReport.distinct('city', { city: { $ne: null } })
        ]);

        res.json({
            success: true,
            data: {
                total_users: usersCount,
                total_reports: reportsCount,
                verified_reports: verifiedCount,
                cities_covered: citiesAgg.length
            }
        });
    } catch (error) {
        console.error('Stats error:', error);
        res.status(500).json({ success: false, message: 'Server error fetching stats' });
    }
};

File: server/controllers/user_controller.py
from fastapi import HTTPException
from bson import ObjectId
from config.db import get_db
from utils.helpers import serialize_doc

async def get_leaderboard(limit: int = 20):
    db = get_db()
    users_cursor = db["users"].find(
        {},
        {"name": 1, "avatar_url": 1, "contribution_score": 1, "reports_count": 1, "verified_reports": 1, "badges": 1, "created_at": 1}
    ).sort("contribution_score", -1).limit(limit)
    
    users = await users_cursor.to_list(length=limit)
    
    leaderboard = []
    for index, user in enumerate(users):
        user_serialized = serialize_doc(user)
        user_serialized["rank"] = index + 1
        user_serialized["badges"] = user_serialized.get("badges") or []
        leaderboard.append(user_serialized)
        
    return {"success": True, "data": leaderboard}

async def get_user_profile(user_id: str):
    db = get_db()
    try:
        obj_id = ObjectId(user_id)
    except Exception:
        raise HTTPException(status_code=400, detail="Invalid user ID format")
        
    user = await db["users"].find_one(
        {"_id": obj_id},
        {"name": 1, "avatar_url": 1, "contribution_score": 1, "reports_count": 1, "verified_reports": 1, "badges": 1, "created_at": 1}
    )
    if not user:
        raise HTTPException(status_code=404, detail="User not found")
        
    reports_cursor = db["pollutionreports"].find(
        {"user_id": obj_id},
        {"title": 1, "category": 1, "severity": 1, "ai_verified": 1, "status": 1, "upvotes": 1, "downvotes": 1, "created_at": 1, "latitude": 1, "longitude": 1}
    ).sort("created_at", -1).limit(10)
    
    reports = await reports_cursor.to_list(length=10)
    
    return {
        "success": True,
        "data": {
            "user": serialize_doc(user),
            "recent_reports": serialize_doc(reports)
        }
    }

async def get_platform_stats():
    db = get_db()
    total_users = await db["users"].count_documents({})
    total_reports = await db["pollutionreports"].count_documents({})
    verified_reports = await db["pollutionreports"].count_documents({"ai_verified": True})
    
    # distinct cities excluding None
    distinct_cities = await db["pollutionreports"].distinct("city", {"city": {"$ne": None}})
    
    return {
        "success": True,
        "data": {
            "total_users": total_users,
            "total_reports": total_reports,
            "verified_reports": verified_reports,
            "cities_covered": len(distinct_cities)
        }
    }
